src/tactic/compute_degree.lean

```diff
@@ -172,8 +298,182 @@ by compute_degree_le
+  resolve_coeff
+| e := skip
+end
+
+namespace interactive
+open compute_degree
+
+/--
+`simp_coeff` assumes that the target is either of the form `f.coeff n = x` or of the form
+`f.coeff n ≠ x`.  It then proceeds to simplify `f.coeff n` recurring to the pair
+`get_lead_coeff/resolve_coeff` to scan the expression of `f` and producing a hopefully simpler
+expression.  After this, it calls on `simp only [a few lemmas]` and `norm_num` to simplify further.
+-/
+meta def simp_coeff : tactic unit :=
+do t ← target >>= instantiate_mvars,
+  (t_is_eq, f, m) ← match t with
+    | `(coeff %%f %%m = _) := return (tt, f, m)
+    | `(coeff %%f %%m ≠ _) := return (ff, f, m)
+    | _ := fail "Goal is not of the form `f.coeff n = x` or `f.coeff n ≠ x`"
+    end,
+  d_nat ← guess_degree' f,
+  m_nat ← eval_expr' ℕ m,
+  guard (d_nat = m_nat) <|> fail!(
+  "`simp_coeff` checks that the expected degree is equal to the degree appearing in `coeff`\n" ++
+  "the expected degree is `{d_nat}`, but you are asking about the coefficient of degree `{m_nat}`"),
+  `(@polynomial %%R %%_) ← infer_type f,
+  c ← mk_instance_cache R,
+  (c, lc) ← get_lead_coeff c f,
+  nn ← get_unused_name "c_c",
+  cf ← to_expr ``(coeff : polynomial %%R → ℕ → %%R),
+  co_eq_co ← mk_app `eq [cf.mk_app [f, m], lc],
+  c_c ← assert nn co_eq_co,
+  interactive.swap,
+  if t_is_eq then refine ``(eq.trans %%c_c _) else refine ``(ne_of_eq_of_ne %%c_c _),
+  try $ tactic.clear c_c,
+  interactive.swap,
+  resolve_coeff
+
+section parsing
+setup_tactic_parser
+
+/--
+`reduce_coeff` takes a polynomial `f` as an input and produces a hypothesis of the form
+`c_c : f.coeff n = <simplified expression>`, where the simplified expression is obtained via the
+`get_lead_coeff/resolve_coeff` pair.
+-/
+meta def reduce_coeff (fp : parse texpr) : tactic unit :=
+do
+  f ← to_expr ``(%%fp) tt ff,
+  exp_deg ← guess_degree' f,
+  `(@polynomial %%R %%_) ← infer_type f,
+  c ← mk_instance_cache R,
+  (c, lc) ← get_lead_coeff c f,
+  nn ← get_unused_name "c_c",
+  cf ← to_expr ``(coeff : polynomial %%R → ℕ → %%R),
+  co_eq_co ← mk_app `eq [cf.mk_app [f, `(exp_deg)], lc],
+  assert nn co_eq_co,
+  resolve_coeff
+
+end parsing
+
+/--  `compute_degree` tries to close goals of the form `f.(nat_)degree = d`.  It converts the
+goal to showing that
+* the degree is at most `d`, calling `compute_degree_le` to solve this case;
+* the coefficient of degree `n` is non-zero, calling `simp_coeff` to simplify this goal.
+Unless the polynomial is particularly complicated, `compute_degree` with either succeed of leave
+a simpler goal to prove.
+ -/
+meta def compute_degree : tactic unit := focus $ do
+t ← target >>= (λ f, whnf f reducible),
+match t with
+  -- the `degree` match implicitly assumes that the `nat_degree` is strictly positive
+| `(    degree %%_ = %%_) := refine ``((degree_eq_iff_nat_degree_eq_of_pos _).mpr _) >> rotate
+| `(nat_degree %%_ = %%_) := single_term_suggestions
```

Comment:
I understand why you want to direct users to specific lemmas if possible, but I wouldn't want the tactic to always fail. An example might be that you are given a list of polynomials `[X^2, X^2 - C a * X - C b, X^2 + C b, 2 * C^2, -X^2]` and want to verify they are all quadratic. Then the proof `simp only [list.forall_mem_cons, list.forall_mem_nil]; split; compute_degree` would be much nicer than `simp only [list.forall_mem_cons, list.forall_mem_nil]; split; try { apply nat_degree_X_pow }; try { apply nat_degree_C_mul_X_pow }; compute_degree` or something similar.

A compromise would be to have a variant along the lines of `compute_degree!` which skips the `single_term_suggestions` check.

Reply:
Now `compute_degree` has a `!` flag.  I also added a test for this.